install: allow instantiated units to be enabled via presets

peterbaouoft opened the pull request:
This patch implements https://github.com/systemd/systemd/issues/9421.

The .preset file now is able to take a rule in the format of:(e.g)
enable foo@.service bar0 bar1 bar2

In the above example, when preset-all is called, all three instances of
foo@bar0.service, foo@bar1.service and foo@bar2.service will be enabled.

When preset is called on a single service(e.g: foo@bar1.service), only
the mentioned one(foo@bar1.service) will be enabled.

peterbaouoft commented:
**NOTE:** This is my first time contributing together with @Bubblemelon. So there may be things that we completely missed, please let us know what we can improve. 

Many thanks for your time and patience!

keszybz commented on src/shared/install.c:
```diff
@@ -2886,10 +2941,20 @@ static int query_presets(const char *name, const Presets presets) {
         case PRESET_ENABLE:
-                log_debug("Preset files say enable %s.", name);
+                if (instance_name_list && *instance_name_list)
+                        STRV_FOREACH(s, *instance_name_list) {
```
Don't put {} around single-line code blocks, see CODING_STYLE.

keszybz commented on src/shared/install.c:
```diff
@@ -2979,22 +3045,49 @@ static int preset_prepare_one(
-                r = install_info_discover(scope, minus, paths, name, SEARCH_FOLLOW_CONFIG_SYMLINKS,
-                                          &i, changes, n_changes);
+                                r = install_info_may_process(i, paths, changes, n_changes);
+                                if (r < 0)
```
Maybe add a short helper func that would combine `install_info_discover` and `install_info_may_process` and use that in both branches here.

peterbaouoft replied:
Thanks, I was considering adding that, but I saw the similar pattern (`install_info_discover` followed by `install_info_may_process`) showed up multiple times in the file. 

Do I create a helper function replacing all of the similar patterns, or would it be sufficient to replace the current ones?

keszybz replied:
Preferably all. If there's more than one or two, please make this a separate patch.

peterbaouoft replied:
Sure, I will be drafting another patch shortly :).

peterbaouoft replied:
Hi, I attempted to create a PR in https://github.com/systemd/systemd/pull/9949, let me know if it is done correctly when you have time :), thanks.

keszybz commented on src/shared/install.c:
```diff
@@ -2979,22 +3045,49 @@ static int preset_prepare_one(
+                                        return r;
+                        }
+                }
+                else {
```
`} else {`, see CODING_STYLE.

keszybz commented on src/shared/install.c:
```diff
@@ -2886,10 +2941,20 @@ static int query_presets(const char *name, const Presets presets) {
-                log_debug("Preset files say disable %s.", name);
+                if (instance_name_list && *instance_name_list)
+                        STRV_FOREACH(s, *instance_name_list) {
+                                log_debug ("Preset files say disable %s.", *s);
```
No {}, no space between func name and paren, see CODING_STYLE.

So... here you use `if (instance_name_list && *instance_name_list)` as the condition, but below just `if (instance_name_list)`. It should be the same in both cases. `if (instance_name_list)` is enough.

peterbaouoft replied:
Hi, yea, I did not notice the {} there, thanks for pointing it out.

For the if statement, I felt the usage here is slightly different to the usage in other condition. The reason why I checked both `instance_name_list` and `*instance_name_list` here is because `instance_name_list` is a triple pointer passed into the function (to assign a `char **` to it) 

Would changes still be needed if that's the case?

keszybz replied:
Ah, OK.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+        const char *parameter;
+        char **iter;
+        _cleanup_free_ char *templated_name = NULL, *prefix = NULL, *instance_name = NULL;
+        int ret;
```
We use `r` for this, please do that for consistency.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+        int ret;
+
+        if (!preset_array || !unit_name_is_valid(unit_name, UNIT_NAME_TEMPLATE))
+                return -1;
```
The return convention should be: 1 – matched, 0 – not matched, < 0 – an errno-style error. Don't return -1, because that means "permission denied".

But since this is a static function, an `assert` is more appropriate then returning an error.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+        if (ret < 0)
+                return ret;
+        parameter = first_word(pattern, templated_name);
+
```
Don't put an empty line between an operation and the check for that operation.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+                return ret;
+
+        /* If the unit name is found in the specified multiple instances, return matching with a null list */
+        if(strv_find(l, instance_name))
```
Always a space between `if` and whatever follows. See CODING_STYLE.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+
+                _cleanup_strv_free_ char **out_strv = NULL;
+                STRV_FOREACH(iter, l) {
+                        _cleanup_free_ char *test;
```
This needs to initalized `= NULL`, because otherwise the uninitialized pointer may be freed on error.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2868,70 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+                        ret = unit_name_build(prefix, *iter, ".service", &test);
+                        if (ret < 0)
+                                return ret;
+                        strv_extend(&out_strv, test);
```
`strv_extend` can return an error too.

peterbaouoft commented:
Hi, thanks for providing detailed review comments and design change! I will be finishing addressing the comments and the design change shortly(likely by tmr). I had also faced a question during implementing the design change. It would be great if you can provide some clarifications :).

 For the `instances`, what's the format of it going to be? I was thinking the following:
- `instances` will be NULL when preset rule, does not follow the pattern of enabling multiple instances. (This also includes `enable foo@.service`,since not multiple instances are specified)

- When multiple instances do get specified. The `instances` will contain all the created service names. E.g: `enable foo@.service a b c` will result in an array of `foo@a.service, foo@b.service, foo@c.service`. 

Does that match what you are thinking?  Thanks for your time.

keszybz commented:
`instances` should be a `char**`, and should contain "a", "b", "c", just to make things simple. You can always construct the full service name when it is really necessary.

peterbaouoft commented:
Thanks for the reply. I force pushed a new commit and addressed the comments and implemented design change. 

Below is the list of changes other than those:
1: The logic for `pattern_match_multiple_instances` is reworked a little. 

Now, both cases for preset multiple instances should be covered. For instance, for a preset rule `disable foo@.service a b c` :
- systemctl preset foo@a.service should disable foo@a.service
- systemctl preset-all should disable `foo@a.service, foo@b.service and foo@c.service` 

2: Since instances field is added for both `disable` and `enable`, tests of using preset rule to disable multiple instances is also added.

Let me know if I interpret the changes needed correctly when you have time :). Thanks for your time and patience!

peterbaouoft commented:
Found a few errors while playing around with the patch... will push a new fix shortly :).

keszybz commented:
Dunno, why would 'disable' with instances be useful? Is seems to make semantics quite complicated, because up the first matching line was enough to decide. If we allow disabling by instance name, some lines become subtractive.

keszybz commented on src/shared/install.c:
```diff
@@ -2825,28 +2849,40 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+                                        return r;
+
                                 rule = (PresetRule) {
                                         .pattern = pattern,
```
Don't store the whole pattern here, just the unit name (and rename the field accordingly). We like to always have a "single source of truth". So there is not need to keep the original pattern, which includes the unit name and the instance names, after it has already been split.

keszybz commented on src/shared/install.c:
```diff
@@ -2979,22 +3082,48 @@ static int preset_prepare_one(
+                                        return r;
+                        }
+                }
+                else
```
Please don't put `else` on a separate line.

peterbaouoft commented:
> Dunno, why would 'disable' with instances be useful? Is seems to make semantics quite complicated, because up the first matching line was enough to decide. If we allow disabling by instance name, some lines become subtractive.

Ah, you are right. I think it was a misunderstanding about requirements/proposal on my side. I will remove the logic for disabling instance names in my next update :). Thanks.

keszybz commented on src/shared/install.c:
```diff
@@ -2825,28 +2849,40 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
                                 if (!pattern)
                                         return -ENOMEM;
 
+                                r = split_pattern_into_instances(pattern, &instances);
```
It'd just throw a warning here ("Invalid line ..., ignoring.") here instead.

peterbaouoft commented:
Hi, thanks for the review, tried to address the comments in the latest commit :). There are also 2 minor changes I made in addition to that, and that was explained in the comments below. 

Thanks for your time and patience.

keszybz commented on src/shared/install.c:
```diff
@@ -2755,6 +2758,37 @@ int unit_file_exists(UnitFileScope scope, const LookupPaths *paths, const char *
+                if (!instances)
+                        return -ENOMEM;
+                if (out_instances)
+                        *out_instances = TAKE_PTR(instances);
```
I'd prefer to see
```C
assert(out_instances);
assert(out_unit_name);
```
at the top of the function instead. There's just one caller, so making this optional doesn't gain us anything.

keszybz commented on src/shared/install.c:
```diff
@@ -2868,15 +2908,73 @@ static int read_presets(UnitFileScope scope, const char *root_dir, Presets *pres
+        r = unit_name_template(unit_name, &templated_name);
+        if (r < 0)
+                return r;
+        parameter = first_word(rule.pattern, templated_name);
```
I'm confused now. Isn't `rule.pattern` a single word now?

peterbaouoft replied:
Sorry, that was a mistake from my side, will change it to `streq` instead. Thanks for pointing that out :).

peterbaouoft commented:
Hi, thanks for your time and review, force pushed another commit addressing the comments. Also cherry-picked #9949 into the second commit.  Let me know what you think when you have time, thanks :).

keszybz commented on src/shared/install.c:
```diff
@@ -2886,10 +2983,18 @@ static int query_presets(const char *name, const Presets presets) {
-                log_debug("Preset files say enable %s.", name);
+                if (instance_name_list && *instance_name_list)
+                        STRV_FOREACH(s, *instance_name_list)
+                                log_debug ("Preset files say enable %s.", *s);
```
No space between the function name and "(".

keszybz commented on src/shared/install.c:
```diff
@@ -1676,6 +1676,28 @@ static int install_info_discover(
+        if (!ret)
+                return install_info_may_process(NULL, paths, changes, n_changes);
+
+        return install_info_may_process(*ret, paths, changes, n_changes);
```
What about collapsing those two cases with `ret ? *ret : NULL` ?

peterbaouoft commented:
Thanks a lot for the fast reviews, force pushed another commit addressing the comments. In addition to that, I also spotted some of the logic for disabling with instances are not removed, removed them and explained below.

peterbaouoft commented on src/shared/install.c:
```diff
@@ -2979,19 +3088,26 @@ static int preset_prepare_one(
-                if (r < 0)
-                        return r;
         } else
                 r = install_info_discover(scope, minus, paths, name, SEARCH_FOLLOW_CONFIG_SYMLINKS,
```
This part remains the same for `r = 0` (which is the disable case), as discussed earlier.

peterbaouoft commented on src/shared/install.c:
```diff
@@ -2886,7 +2990,11 @@ static int query_presets(const char *name, const Presets presets) {
+                        log_debug("Preset files say enable %s.", name);
                 return 1;
         case PRESET_DISABLE:
                 log_debug("Preset files say disable %s.", name);
```
Similarly, we don't have instance name list for disable case. I felt therefore, no looping is required here :).

peterbaouoft commented:
@keszybz, Thanks a lot for guiding me and @Bubblemelon along!

ajeddeloh commented:
I'm not sure about how to approach adding support for this in Ignition itself. Probably via a compile time switch like "systemdversion=240" which lets us enable things only supported in 240+. We can't just enable it because then it will write out invalid files for anything that doesn't have systemd new enough.
